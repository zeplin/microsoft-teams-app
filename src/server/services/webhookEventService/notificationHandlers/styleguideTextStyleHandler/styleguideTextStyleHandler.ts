import {
    StyleguideTextStyleCreateEvent,
    StyleguideTextStyleUpdateEvent,
    WebhookEvent
} from "../../../../adapters/zeplin/types";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";

type Event = StyleguideTextStyleCreateEvent | StyleguideTextStyleUpdateEvent;

class StyleguideTextStyleHandler extends NotificationHandler<Event> {
    delay = SHORT_DELAY;
    private getText(events: Event[]): string {
        const [{
            payload: {
                action,
                context: {
                    styleguide: {
                        name: styleguideName
                    }
                },
                resource: {
                    data: {
                        name: pivotTextStyleName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added to" : "updated in";
        return events.length === 1
            ? md`**${pivotTextStyleName}** is ${actionText} _${styleguideName}_! ${getRandomEmoji()}`
            : md`**${events.length} text styles** are ${actionText} _${styleguideName}_! ${getRandomEmoji()}`;
    }

    private getWebappURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const pathname = `styleguide/${styleguideId}/textstyles`;
        const searchParams = {
            tsid: events.map(event => event.payload.resource.id)
        };

        return getWebAppURL(pathname, searchParams);
    }

    private getZeplinAppURI(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const searchParams = {
            stid: styleguideId,
            tsids: events.map(event => event.payload.resource.id)
        };

        return getRedirectURLForZeplinApp("textStyles", searchParams);
    }

    shouldHandleEvent(event: WebhookEvent): event is Event {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: Event[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Make sure your stylesheets are up to date!"
            },
            links: [{
                title: "Open in App",
                url: this.getZeplinAppURI(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const styleguideTextStyleHandler = new StyleguideTextStyleHandler();
