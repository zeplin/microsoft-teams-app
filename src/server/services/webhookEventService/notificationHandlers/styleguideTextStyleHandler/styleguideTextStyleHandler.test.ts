import {
    Styleguide,
    StyleguideTextStyleCreatedEvent,
    StyleguideTextStyleUpdatedEvent
} from "@zeplin/sdk";

import { styleguideTextStyleHandler } from "./styleguideTextStyleHandler";

type GetDummyEventParams = {
    action?: string;
    textStyleId?: string;
    textStyleName?: string;
    styleguidePlatform?: Styleguide["platform"];
}

function getDummyEvent({
    action = "created",
    textStyleId = "textStyleId",
    textStyleName = "textStyleName",
    styleguidePlatform = "web"
}: GetDummyEventParams = {}): StyleguideTextStyleCreatedEvent | StyleguideTextStyleUpdatedEvent {
    return {
        action,
        context: {
            styleguide: {
                id: "styleguideId",
                name: "styleguideName",
                platform: styleguidePlatform
            }
        },
        resource: {
            id: textStyleId,
            data: {
                id: textStyleId,
                name: textStyleName
            }
        }
    } as StyleguideTextStyleCreatedEvent | StyleguideTextStyleUpdatedEvent;
}

describe("styleguideTextStyleHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideTextStyleHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideTextStyleHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Styleguide["platform"]>([
                "android",
                "ios",
                "macos",
                "web",
                "base"
            ])("should match snapshot when styleguide platform is %s",
                styleguidePlatform => {
                    expect(
                        styleguideTextStyleHandler.getTeamsMessage([getDummyEvent({ styleguidePlatform })])
                    ).toMatchSnapshot();
                }
            );
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideTextStyleHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideTextStyleHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Styleguide["platform"]>([
                "android",
                "ios",
                "macos",
                "web",
                "base"
            ])("should match snapshot when styleguide platform is %s",
                styleguidePlatform => {
                    expect(
                        styleguideTextStyleHandler.getTeamsMessage([getDummyEvent({ styleguidePlatform })])
                    ).toMatchSnapshot();
                }
            );
        });
    });
});
