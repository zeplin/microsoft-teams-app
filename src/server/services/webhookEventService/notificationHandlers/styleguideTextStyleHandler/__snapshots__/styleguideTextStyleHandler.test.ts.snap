// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styleguideTextStyleHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 text styles** are added in _styleguideName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fstid%3DstyleguideId%26tsids%3DtextStyleId%2CanotherTextStyleId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added in _styleguideName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fstid%3DstyleguideId%26tsids%3DtextStyleId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 text styles** are updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fstid%3DstyleguideId%26tsids%3DtextStyleId%2CanotherTextStyleId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideTextStyleHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fstid%3DstyleguideId%26tsids%3DtextStyleId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/styleguide/styleguideId/styleguide/textstyles?tsid=textStyleId"
        }
      ]
    }
  ]
}
`;
